File: scripts/count_genome.rb
#!/usr/bin/env ruby

GENOME_FILE = '/tmp/genome.txt'

STATUS = {
  ok: 0,
  error: 1,
}
$queues = []

def main()
  exit_status = STATUS[:ok]

  begin
    open(GENOME_FILE) { |file|
      # Implement as a map-reduce style intentionally.
      while line = file.gets
        line.chomp!
        map_obj = map_for_line(line)

        queue = {
          key: line,
          value: map_obj[:value],
        }
        $queues.push(queue)
      end

      total_map = {}
      while $queues.size > 0
        queue = $queues.shift
        total_map = reduce!(total_map, queue[:value])
      end

      print_counted_map(total_map)
    }
  rescue RuntimeError => e
    puts e
    exit_status = STATUS[:error]
  end
end

def map_for_line(line)
  counted_map = {}

  if line.nil?
    raise ArgumentError.new("line is required.")
  end
  line.each_char { |c|
    if !counted_map.key?(c)
      counted_map[c] = 0
    end
    counted_map[c] += 1
  }

  { key: line,
    value: counted_map,
  }
end

def reduce!(total_map, map)
  if total_map.nil? or map.nil?
    raise ArgumentError.new("total_map or map is required.")
  end

  map.each_pair { |c, count|
    if !total_map.key?(c)
        total_map[c] = 0
    end
    total_map[c] += count;
  }

  total_map
end

def print_counted_map(counted_map)
  if counted_map.nil?
    raise ArgumentError.new("counted_map is required.")
  end

  total = 0
  sorted_keys = counted_map.each_key
    .sort
  sorted_keys.each { |c|
    count = counted_map[c]
    puts "#{c}: #{count}\n"
    total += count
  }
  puts "Total: #{total}\n"
end

main

exit(0)

File: scripts/count_genome.pl
#!/usr/bin/env perl

use strict;
use warnings;
use Carp;
use English qw( -no_match_vars );
use Try::Tiny;

my $STATUS = +{
    ok    => 0,
    error => 1,
};

my $GENOME_FILE = '/tmp/genome.txt';
my $_queues_ref = [];

sub main {
    my $exit_status = $STATUS->{ok};
    try {
        my $fh = undef;
        try {
            open $fh, '<', $GENOME_FILE
                or croak "$GENOME_FILE $OS_ERROR";

            # Implement as a map-reduce style intentionally.
            while (my $line = <$fh>) {
                chomp $line;
                my ($line, $counted_map) = map_for_line($line);
                push @{$_queues_ref}, +{
                    key => $line,
                    value => $counted_map,
                };
            }

            my $total_map = +{};
            while (scalar @{$_queues_ref} > 0) {
                my $queue = shift @{$_queues_ref};
                $total_map = reduce($total_map, $queue->{value});
            }

            print_counted_map($total_map);
        }
        finally {
            if ( defined $fh ) {
                close $fh;
            }
        };
    }
    catch {
        $exit_status = $STATUS->{error};
    };

    return $exit_status;
}

sub map_for_line {
    my ($line) = @_;

    my $counted_map = +{};

    if ( !defined $line ) {
        croak 'line is required.';
    }
    for (my $pos = 0; $pos < length($line); $pos++) {
        my $c = substr($line, $pos, 1);
        $counted_map->{$c}++;
    }

    # map's key, value
    return ($line, $counted_map);
}

sub reduce {
    my ($total_map, $map) = @_;

    if ( !defined $total_map || !defined $map ) {
        croak 'total_map or map is required.';
    }

    for my $c (keys %{$map}) {
        $total_map->{$c} += $map->{$c};
    }

    return $total_map;
}

sub print_counted_map {
    my ($counted_map) = @_;

    my $total = 0;
    for my $c (sort(keys %{$counted_map})) {
        printf "%s: %d\n", $c, $counted_map->{$c};
        $total += $counted_map->{$c};
    }
    printf "Total: %d\n", $total;

    return;
}

my $exit_status = main();
exit $exit_status;

